homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,432 @@
+    def async_save(self) -> None:
+        """Save the schedules to local storage."""
+
+        def save() -> None:
+            """Save."""
+            save_json(
+                self._hass.config.path(PERSISTENCE),
+                {
+                    schedule_id: self.as_dict(schedule_id)
+                    for schedule_id, schedule in self.schedules.items()
+                },
+            )
+
+        self._hass.async_add_job(save)
+        self._hass.bus.async_fire(EVENT)
+
+    @callback
+    def async_schedule_next_instance(
+        self, schedule_id: str, *, starting_from: datetime = None
+    ) -> bool:
+        """Schedule the next instance of a schedule."""
+        schedule = self.schedules[schedule_id]
+
+        if not starting_from:
+            starting_from = datetime.now()
+        start_dt = schedule[CONF_RRULE].after(starting_from, inc=True)
+
+        # If no recurrence rule exists or we've run out of recurrences, delete the
+        # schedule and return:
+        if not start_dt or not schedule.get(CONF_RRULE):
+            self.async_delete(schedule_id, remove_listeners=False)
+            return
+
+        if schedule.get(CONF_DURATION):
+            end_dt = start_dt + timedelta(seconds=schedule[CONF_DURATION])
+        else:
+            end_dt = None
+
+        instance = ScheduleInstance(
+            self._hass, schedule[CONF_ENTITY_ID], schedule[CONF_ACTIVATION_CONTEXT]
+        )
+
+        @callback
+        def schedule_start(call: datetime) -> None:
+            """Trigger when the schedule starts."""
+            instance.async_trigger_scene()
+            if not end_dt:
+                self.async_schedule_next_instance(schedule_id)
+
+        @callback
+        def schedule_end(call: datetime) -> None:
+            """Trigger when the schedule ends."""
+            instance.async_undo_scene()
+            self.async_schedule_next_instance(schedule_id)
+
+        self._async_activation_listeners[schedule_id] = async_track_point_in_time(
+            self._hass, schedule_start, start_dt
+        )
+        if end_dt:
+            self._async_deactivation_listeners[schedule_id] = async_track_point_in_time(
+                self._hass, schedule_end, end_dt
+            )
+
+        _LOGGER.info(
+            'Scheduled instance of schedule "%s" (start: %s / end: %s)',
+            schedule_id,
+            start_dt,
+            end_dt,
+        )
+
+    def async_update(self, schedule_id: str, new_data: dict) -> dict:
+        """Update a schedule."""
+        if schedule_id not in self.schedules:
+            raise KeyError
+
+        new_data = SCHEDULE_UPDATE_SCHEMA(new_data)
+        data = self.async_delete(schedule_id, save=False)
+        data.update(new_data)
+        return self.async_create(
+            data[CONF_ENTITY_ID],
+            data[CONF_RRULE],
+            duration=data[CONF_DURATION],
+            activation_context_id=data[CONF_ACTIVATION_CONTEXT].id,
+        )
+
+
+@callback
```

Comment:
All these things need to become async and only return when the creation is done.